Add functions to apply forces given a global position instead of an origin offset

GustJc opened the pull request:
# What problem(s) does this PR solve?

- Closes https://github.com/godotengine/godot-proposals/issues/12115

Adds physics methods for `apply_force_at_position(force, world_pos)` and `apply_impulse_at_position(force, world_pos)`.

Which uses a **global coordinate position** to say where the force is applied.
They represent a more common use-case where you want to apply a force to a global position.

For example, you can apply the force at a given node's position that is attached to the player's body without having to use subtract the rigidbodies global_position.
So _apply_force(force, (global_point - global_position) )_ == _apply_force_at_position(force, global_point)_

https://github.com/user-attachments/assets/3e2d2628-c6a8-4ac6-a94f-71996c0b7dfe

https://github.com/user-attachments/assets/2d0e2358-0023-4c79-b72b-fee2048c6d04


# Additional information

The **reason they are being added** is to not change/break the behavior of the already in use `apply_force` and `apply_impulse` which can take an optional global offset value for the position.
So this PR **will not break compatibility.**

Using them, it is now possible to apply forces at child objects position's without any hassle
```gdscript
apply_force_at_position(-%ForceLeft.global_basis.y * FORCE, %ForceLeft.global_position)
```
Previously, to get the same behavior you'd need to use
```gdscript
var body_offset := %ForceLeft.global_position - global_position
apply_force(-%ForceLeft.global_basis.y * FORCE, body_offset)
```
Which is **more confusing** and can lead to a lot of errors if you don't understand what a global offset means.

_(My first time using godot, years ago, I read it wrong and thought it was a local offset in the same sense as a local position, so the offset would be relative to the body's rotation. Made to lose two days thinking my math was wrong and redoing stuff)_

---
I've made a test MRP you can play around to check if everything is working as it should. 
From my simple tests, it seems to be okay.
[PR_MRP_physics_apply_at_position.zip](https://github.com/user-attachments/files/19560234/PR_MRP_physics_apply_at_position.zip)

Applying forces at the red cubes when pressing left/right and space 

https://github.com/user-attachments/assets/3477e1ab-c4dd-4bf1-9d5c-0951c4bbf45b

---
For reference, only Godot seems to use this global offset approach. 
With all other engines I've researched using global coordinates instead. 
This also include Jolt's own internals with AddForce(force, global_coordinate)

Edit: Force pushed changes, Jolt parameters now casts to JPH::RVec3 to not break tests.

GustJc commented:
> If we do decide to go down this route, I would personally prefer to see that the existing methods be "renamed" to `apply_force_at_offset`, `apply_impulse_at_offset` and `add_constant_force_at_offset`, while leaving the existing methods in there, but marked as deprecated. That way it's still not a breaking chance (right now) and we can clearly signal the upcoming breakage for however long we feel is necessary.

That's a great idea, it prepares the changes for the future (Godot 5) and alert people of the new functions.

> Also, you should ideally change this on the 2D side of things as well, which similarly takes an offset currently.

In 2D the problem is that big but I agree that it should have the 2D counterpart to keep things in sync.
I also forgot that `add_constant_force` takes an offset, so I should create a `add_constant_force_at_position`.

I'll **wait for a while to see how people few about this change** and implement the changes if they are okay with this.
Heres a list of changes just keep track later:
- [ ] Add `...at_position` for 2D
- [ ] Add `constant_force_at_position`
- [ ] Add the deprecated tag to `apply_force/impulse/constant_force`
- [ ] Add renamed `..._at_offset` for `apply_force/impulse/constant_force`

mihe commented:
> I also forgot that `add_constant_force` takes an offset, but its behavior is a bit different than the apply_force/impulse.
> Using a global fixed position for this one would be strange, so I'm not sure if I should also add an `add_constant_force_at_position` or not.

I don't think a `add_constant_force_at_position` makes much sense, no.

GustJc committed:
Add apply_force_at_position and impulse for global_position instead of offset

GustJc commented:
Rebased to remove conflicts. 

And added the parity functions for the 2D physics now. 
Now this **should be complete, and can hopefully be considered to be merged now.**

I don't think we should add the deprecated tag to apply_force/impulse since it seems to be the preferred way to apply central forces still, and it would cause confusion.

Having these  _at_position global alternatives should already solve most of the frustration with using positional forces that users coming to Godot seems to experience due to misunderstanding the local offset position.
